frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSquashAction.java

```diff
@@ -0,0 +1,173 @@
+        .filter(b -> b.isNonRoot())
+        .map(b -> {
+          assert b.isNonRoot() : "branch is root";
+          return b.asNonRoot();
+        });
+    val commitsOption = nonRootBranchOption
+        .map(b -> b.getCommits());
+    val parent = nonRootBranchOption
+        .flatMap(b -> b.getForkPoint()).getOrNull();
+    val syncToParentStatus = nonRootBranchOption.map(b -> b.getSyncToParentStatus()).getOrNull();
+    val gitRepository = getSelectedGitRepository(anActionEvent).getOrNull();
+    val branchName = branchNameOption.getOrNull();
+    val commits = commitsOption.getOrNull();
+
+    if (commits != null && gitRepository != null && parent != null && branchName != null
+        && syncToParentStatus != InSyncButForkPointOff) {
+      val isSquashingCurrent = Option.of(gitRepository.getCurrentBranch()).map(b -> b.getName())
+          .flatMap(cbn -> branchNameOption.map(bn -> bn.equals(cbn))).getOrElse(false);
+      doSquash(project, gitRepository, parent, commits, branchName, isSquashingCurrent);
+    }
+  }
+
+  @UIEffect
+  private void doSquash(Project project,
+      GitRepository gitRepository,
+      ICommitOfManagedBranch parent,
+      List<ICommitOfManagedBranch> commits,
+      String branchName,
+      boolean isSquashingCurrent) {
+    val vcsCommitMetadataAndMessage = commits.foldLeft(
+        new Tuple2<List<VcsCommitMetadata>, String>(List.empty(), ""),
+        (acc, commit) -> new Tuple2<>(
+            acc._1.append(new VcsCommitMetadataAdapter(parent, commit)), "${acc._2}${commit.getFullMessage()}${LS}${LS}"));
+
+    val dialog = new GitNewCommitMessageActionDialog(
+        /* project */ project,
+        /* message */ vcsCommitMetadataAndMessage._2,
+        /* title */ getNonHtmlString("action.GitMachete.BaseSquashAction.dialog.title"),
+        /* dialogLabel */ getNonHtmlString("action.GitMachete.BaseSquashAction.dialog.label"));
+
+    String taskName = isSquashingCurrent
+        ? getString("action.GitMachete.BaseSquashAction.task-title.current")
+        : getString("action.GitMachete.BaseSquashAction.task-title.non-current");
+
+    dialog.show(
+        newMessage -> {
+          new Task.Backgroundable(project, taskName) {
+            @Override
+            @UIThreadUnsafe
+            public void run(ProgressIndicator indicator) {
+              LOG.info("Checking out '${branchName}' branch and squashing it");
+
+              if (!isSquashingCurrent) {
+                CheckoutSelectedBranchAction.doCheckout(project, indicator, branchName, gitRepository);
+              }
+
+              val commitsToSquash = vcsCommitMetadataAndMessage._1.toJavaList();
+              val operationResult = new GitSquashOperation(gitRepository).execute(commitsToSquash, newMessage);
+
+              // Hackish approach to check for internal sealed class git4idea.rebase.log.GitCommitEditingOperationResult.Complete.
```

Comment:
`sealed` class... but hmmm, we're still on Java 11 right? 🤔

Reply:
In what sense is the class _sealed_?

Reply:
GitCommitEditingOperationResult is a **kotlin** class

Reply:
Pls clarify that in the comment (as _sealedness_ is independently implemented in Java 17, Scala and Kotlin)